src/main.cpp

```diff
@@ -3460,8 +3470,10 @@ void static CheckBlockIndex()
     CBlockIndex* pindexFirstNotScriptsValid = NULL; // Oldest ancestor of pindex which does not have BLOCK_VALID_SCRIPTS (regardless of being valid or not).
     while (pindex != NULL) {
         nNodes++;
+        // Not calling HaveBlockData() directly, since that can change state.
```

Comment:
This comment points out that the HaveBlockData() name is very misleading. Could you change the name to indicate that it also updates the state?